iris_ugrid/regrid.py

```diff
@@ -0,0 +1,420 @@
+            the start index for describing its elements. This makes no
+            difference to the regridding calculation and will only affect the
+            intermediate ESMF objects, should the user need access to them.
+        * areas
+            either None or a numpy array describing the areas associated with
+            each face. If None, then ESMF will use its own calculated areas.
+        """
+        self.node_coords = node_coords
+        self.fnc = face_node_connectivity
+        self.nsi = node_start_index
+        self.esi = elem_start_index
+        self.areas = areas
+
+    def _as_esmf_info(self):
+        # ESMF uses a slightly different format to UGRID,
+        # the data must be translated into a form ESMF understands
+        num_node = self.node_coords.shape[0]
+        num_elem = self.fnc.shape[0]
+        nodeId = np.array(range(self.nsi, self.nsi + num_node))
+        nodeCoord = self.node_coords.flatten()
+        nodeOwner = np.zeros([num_node])  # regridding currently serial
+        elemId = np.array(range(self.esi, self.esi + num_elem))
+        elemType = self.fnc.count(axis=1)
+        # Experiments seem to indicate that ESMF is using 0 indexing here
+        elemConn = self.fnc.compressed() - self.nsi
+        result = (
+            num_node,
+            num_elem,
+            nodeId,
+            nodeCoord,
+            nodeOwner,
+            elemId,
+            elemType,
+            elemConn,
+            self.areas,
+        )
+        return result
+
+    def _make_esmf_mesh(self):
+        info = self._as_esmf_info()
+        (
+            num_node,
+            num_elem,
+            nodeId,
+            nodeCoord,
+            nodeOwner,
+            elemId,
+            elemType,
+            elemConn,
+            areas,
+        ) = info
+        # ESMF can handle other dimensionalities, but we are unlikely
+        # to make such a use of ESMF
+        emesh = ESMF.Mesh(
+            parametric_dim=2, spatial_dim=2, coord_sys=ESMF.CoordSys.SPH_DEG
+        )
+
+        emesh.add_nodes(num_node, nodeId, nodeCoord, nodeOwner)
+        emesh.add_elements(
+            num_elem, elemId, elemType, elemConn, element_area=areas
+        )
+        return emesh
+
+    def make_esmf_field(self):
+        mesh = self._make_esmf_mesh()
+        field = ESMF.Field(mesh, meshloc=ESMF.MeshLoc.ELEMENT)
+        return field
+
+    def size(self):
+        return self.fnc.shape[0]
+
+    def _index_offset(self):
+        return self.esi
+
+    def _flatten_array(self, array):
+        return array
+
+    def _unflatten_array(self, array):
+        return array
+
+
+class GridInfo:
+    """
+    This class holds information about lat-lon type grids.
+    It contains methods for translating this information into ESMF objects.
+    """
+
+    def __init__(
+        self,
+        lons,
+        lats,
+        lonbounds,
+        latbounds,
+        crs=ccrs.Geodetic(),
```

Comment:
If we provide a crs, then presumably 'lons' and 'lats' are not necessarily lons+lats, but can be more general X and Y coord values.  In which case, it may also be worth making clear that the units of these coords are determined by the crs.

Alternatively, we should probably state that lats + lons are in degrees (which is not a given in Iris/CF).